File: Tercera Entrega/sdi3-37.Cli-REST/src/com/sdi/presentation/user/action/TareasCategoriaAction.java
package com.sdi.presentation.user.action;

import java.text.SimpleDateFormat;
import java.util.List;

import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.core.GenericType;
import javax.ws.rs.core.Response;

import com.sdi.dto.Category;
import com.sdi.dto.Task;
import com.sdi.presentation.user.Authenticator;
import com.sdi.presentation.user.Sesion;

import alb.util.console.Console;
import alb.util.menu.Action;

public class TareasCategoriaAction implements Action {

	public TareasCategoriaAction() {}
	
	@Override
	public void execute() throws Exception {
		Console.println("------TAREAS PENDIENTES Y RETRASADAS DE UNA CATEGORIA------");
		
		//Pedir datos
		String categoryName = Console.readString("Nombre de la categoría");
		
		//Comprobamos que exista la categoria para nuestro usuario
		GenericType<Category> modelo = new GenericType<Category>() {};
		
		Response response = ClientBuilder.newClient()				
				.register(new Authenticator(Sesion.getInstance().getUser().getLogin(), Sesion.getInstance().getUser().getPassword()))
				.target(Sesion.getInstance().getRestServiceUrl())
				.path("buscarCategoria")
				.path("id="+String.valueOf(Sesion.getInstance().getUser().getId())+"&&categoria="+categoryName)
				.request()
				.get();
		
		Category categoria = null;
		if(response.getStatus() == 200) {
			categoria = response.readEntity(modelo);
		}
		else {
			Console.println("\tError en la peticion de comprobar categoria");
		}
		
		//Si existe mostramos las tareas
		if(categoria != null) {
			GenericType<List<Task>> modelo2 = new GenericType<List<Task>>() {};
			
			Response response2 = ClientBuilder.newClient()				
					.register(new Authenticator(Sesion.getInstance().getUser().getLogin(), Sesion.getInstance().getUser().getPassword()))
					.target(Sesion.getInstance().getRestServiceUrl())
					.path("tareas")
					.path("id="+String.valueOf(Sesion.getInstance().getUser().getId())+"&&categoria="+categoryName)
					.request()
					.get();					
			
			List<Task> tareas = null;
			if(response2.getStatus() == 200) {
				tareas = response2.readEntity(modelo2);
			}
			else {
				Console.println("\tError en la peticion de obtener tareas de una categoria");
			}
			print(tareas);
		} 
		else {
			Console.println("\tNo existe la categoría introducida");
		}
		
		Console.println("-----------------------------------------------------------");
		
	}

	private void print(List<Task> tareas) {
		Console.printf("%-20s%-20s%-20s%-25s%-20s\n", "ID", "TITULO", "FECHA_CREACION", "FECHA_FINALIZACION", "FECHA_PLANEADA");
		SimpleDateFormat formateador = new SimpleDateFormat("dd/MM/yyyy");
		for(Task t: tareas) {
			Console.printf("%-20s", t.getId());
			Console.printf("%-20s", t.getTitle());
			Console.printf("%-20s", formateador.format(t.getCreated()));
			Console.printf("%-25s", "");
			Console.printf("%-20s\n", formateador.format(t.getPlanned()));			
		}
	}
}

File: Tercera Entrega/sdi3-37.Cli-REST-C_Sharp/sdi3-37.Cli-REST-C_Sharp/action/TareasCategoriaAction.cs
using sdi3_37.Cli_REST_C_Sharp.dto;
using sdi3_37.Cli_REST_C_Sharp.util;
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;

namespace sdi3_37.Cli_REST_C_Sharp.action
{
    class TareasCategoriaAction : Action
    {
        private static Category category;
        private static String json;       

        public TareasCategoriaAction()
        {

        }

        public void execute()
        {
            Console.WriteLine("------TAREAS PENDIENTES Y RETRASADAS DE UNA CATEGORIA------");

            //Pedir datos
            Console.Write("Nombre de la categoria: ");
            var categoryName = Console.ReadLine();
            Console.WriteLine();

            //Comprobamos que exista la categoria para nuestro usuario
            comprobarCategoria(categoryName).Wait();

            if(category != null)
            {              
                listarTareas(category.name).Wait();
                print(Util.parsearTareas(json));                
            }
            else
            {
                Console.WriteLine("\tNo existe la categoría introducida");
            }
            Console.WriteLine("-----------------------------------------------------------");
        }

        static async Task comprobarCategoria (String categoryName)
        {
            using (var cliente = new HttpClient())
            {
                cliente.BaseAddress = new Uri(Sesion.Instance.URL);
                cliente.DefaultRequestHeaders.Accept.Clear();
                cliente.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

                HttpResponseMessage response = await cliente.GetAsync("buscarCategoria/id="+Sesion.Instance.User.id+"&&categoria="+categoryName);
                if (response.IsSuccessStatusCode)
                {
                    Category categoria = await response.Content.ReadAsAsync<Category>();
                    category = categoria;
                }
            }
        }

        static async Task listarTareas (String categoryName)
        {
            using (var cliente = new HttpClient())
            {
                cliente.BaseAddress = new Uri(Sesion.Instance.URL);
                cliente.DefaultRequestHeaders.Accept.Clear();
                cliente.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

                HttpResponseMessage response = await cliente.GetAsync("tareas/id=" + Sesion.Instance.User.id + "&&categoria=" + categoryName);
                if (response.IsSuccessStatusCode)
                {
                    json = await response.Content.ReadAsStringAsync();                 
                }
            }
        }

        private static void print (List<Tarea> tareas)
        {
            Console.WriteLine(String.Format("{0,-20}{1,-20}{2,-20}{3,-25}{4,-20}", "ID", "TITULO", "FECHA_CREACION", "FECHA_FINALIZACION", "FECHA_PLANEADA"));

            foreach(Tarea t in tareas) {
                Console.Write(String.Format("{0,-20}", t.id));
                Console.Write(String.Format("{0,-20}", t.title));
                Console.Write(String.Format("{0,-20}", t.created.ToString("d")));
                Console.Write(String.Format("{0,-25}", ""));
                Console.WriteLine(String.Format("{0,-20}", t.planned.ToString("d")));
            }
        }
    }
}
